pkg/utils/managedresources/builder/managedresources.go

```diff
@@ -112,28 +115,94 @@ func (m *ManagedResource) DeletePersistentVolumeClaims(v bool) *ManagedResource
 	return m
 }
 
-// Reconcile creates or updates the ManagedResource.
+// Reconcile creates or updates the ManagedResource as well as marks all referenced secrets as garbage collectable.
 func (m *ManagedResource) Reconcile(ctx context.Context) error {
 	resource := &resourcesv1alpha1.ManagedResource{
 		ObjectMeta: metav1.ObjectMeta{Name: m.resource.Name, Namespace: m.resource.Namespace},
 	}
 
-	_, err := controllerutil.CreateOrUpdate(ctx, m.client, resource, func() error {
+	mutateFn := func(obj *resourcesv1alpha1.ManagedResource) {
 		for k, v := range m.labels {
-			metav1.SetMetaDataLabel(&resource.ObjectMeta, k, v)
+			metav1.SetMetaDataLabel(&obj.ObjectMeta, k, v)
 		}
 
 		for k, v := range m.annotations {
-			metav1.SetMetaDataAnnotation(&resource.ObjectMeta, k, v)
+			metav1.SetMetaDataAnnotation(&obj.ObjectMeta, k, v)
 		}
 
-		resource.Spec = m.resource.Spec
+		obj.Spec = m.resource.Spec
+
+		// the annotations should be injected after the spec is updated!
+		utilruntime.Must(references.InjectAnnotations(obj))
+	}
+
+	if err := m.client.Get(ctx, client.ObjectKeyFromObject(resource), resource); err != nil {
```

Comment:
Can't we stick to `controllerutil.CreateOrUpdate` here?

Reply:
We probably can, but I do not find it intuitive to mutate secrets in the `controllerutil.CreateOrUpdate` function called on a managed resource object.